File: 1895-minimum-number-of-operations-to-move-all-balls-to-each-box/1895-minimum-number-of-operations-to-move-all-balls-to-each-box.py
from typing import List

class Solution:
    def minOperations(self, boxes: str) -> List[int]:
        n = len(boxes)
        answer = [0] * n

        # Pass 1: Left to right
        count = 0  # Number of balls encountered
        operations = 0  # Operations needed to move all balls so far
        for i in range(n):
            answer[i] += operations
            if boxes[i] == '1':
                count += 1
            operations += count

        # Pass 2: Right to left
        count = 0
        operations = 0
        for i in range(n - 1, -1, -1):
            answer[i] += operations
            if boxes[i] == '1':
                count += 1
            operations += count

        return answer

File: 1895-minimum-number-of-operations-to-move-all-balls-to-each-box/1895-minimum-number-of-operations-to-move-all-balls-to-each-box.java
class Solution {
    public int[] minOperations(String boxes) {
        int n = boxes.length();
        int ans[] = new int[n];
        int count = 0;
        // for(int i=0;i<n;i++){
        //     count = 0;
        //     for(int j=0;j<n;j++){
        //         if(boxes.charAt(j)=='1') count += Math.abs(j-i);
        //     }
        //     ans[i] = count ;

        // }
        List<Integer> indexOfOnes = new ArrayList<Integer>();
        for(int i=0;i<n;i++){
            if(boxes.charAt(i)=='1'){
                indexOfOnes.add(i);
            }
        }
        for(int i=0;i<n;i++){
            count = 0;
            for(int num:indexOfOnes){
                count += Math.abs(num-i);
            }
            ans[i] = count;

        }
        return ans;
        
    }
}